src/Workspaces/Remote/ServiceHub/Host/RemoteWorkspace.cs

```diff
@@ -159,79 +288,54 @@ public async Task UpdatePrimaryBranchSolutionAsync(AssetProvider assetProvider,
                     // create updated solution off the baseSolution
                     var solution = await updater.CreateSolutionAsync(solutionChecksum).ConfigureAwait(false);
 
-                    if (fromPrimaryBranch)
-                    {
-                        // if the solutionChecksum is for primary branch, update primary workspace cache with the solution
-                        return UpdateSolutionIfPossible(solution, workspaceVersion);
-                    }
-
-                    // otherwise, just return the solution
-                    return solution;
+                    // if the solutionChecksum is for primary branch, update primary workspace cache with the solution
+                    return await UpdateSolutionIfPossibleAsync(
+                        workspaceVersion, fromPrimaryBranch, solution, cancellationToken).ConfigureAwait(false);
                 }
+                else
+                {
+                    // we need new solution. bulk sync all asset for the solution first.
+                    await assetProvider.SynchronizeSolutionAssetsAsync(solutionChecksum, cancellationToken).ConfigureAwait(false);
 
-                // we need new solution. bulk sync all asset for the solution first.
-                await assetProvider.SynchronizeSolutionAssetsAsync(solutionChecksum, cancellationToken).ConfigureAwait(false);
-
-                // get new solution info and options
-                var (solutionInfo, options) = await assetProvider.CreateSolutionInfoAndOptionsAsync(solutionChecksum, cancellationToken).ConfigureAwait(false);
+                    // get new solution info and options
+                    var (solutionInfo, options) = await assetProvider.CreateSolutionInfoAndOptionsAsync(solutionChecksum, cancellationToken).ConfigureAwait(false);
 
-                if (fromPrimaryBranch)
-                {
                     // if the solutionChecksum is for primary branch, update primary workspace cache with new solution
-                    if (TrySetCurrentSolution(solutionInfo, workspaceVersion, options, out var solution))
-                    {
+                    var solution = await TrySetCurrentSolutionAsync(
+                        solutionInfo, workspaceVersion, fromPrimaryBranch, options, cancellationToken).ConfigureAwait(false);
+                    if (solution != null)
                         return solution;
-                    }
-                }
```

Comment:
note: ideall we would not have to set Workspace.CurrentSolution at all (using either UpdateSolutionIfPossible, or TrySetCurrentSolution).  @tmat Do you know how close we are to nothing on the OOP side using Workspace.CurrentSolution and instead having everything just use the solution snapshots they care given?  If we're almost there, i can continue work here and make it so that this updating doesn't even both with that anymore (including potentially making RemoteWorkspace.CurrentSnapshot throw.

Reply:
I don't think we are close. Pls file an issue and link it from here.